### PR TITLE
Add a design doc for decoupling the client code out of SQLFlow Core

### LINKED ISSUE #1744: Proposal: Make repl call SQLFlow server
Currently, the binary repl call [`func RunSQLProgram`](https://github.com/sql-machine-learning/sqlflow/blob/5ab490f740decbb8c3f293461c7920d5ef2f8979/pkg/sql/executor_ir.go#L52) locally.

I propose to 

1. Rename repl into sqlflow.
1. Make repl call SQLFlow server, just like mysql calls MySQL server.
1. Move part of `RunSQLProgram` in `executor_ir.go` to `pkg/server` and the rest to `pkg/step/executor/`.  The former part calls the later part.

BTW, we need `cmd/step`, which calls `pkg/step/executor/` and is called by the generated workflow to run each step, as explained in https://github.com/sql-machine-learning/sqlflow/issues/1746.

A side-effect of this proposal is to reduce and finally remove the file `execute_ir.go`, as the IR is not executable -- it is the submitter program generated from the IR which is executable.

### PR DESCRIPTION
**Pull request by shendiaomo**
Fix https://github.com/sql-machine-learning/sqlflow/issues/1744

### CODE REVIEW

File: doc/design/cli.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
+# SQLFlow Command Line Interface
```
Comment (wangkuiyi) [inline doc/design/cli.md:1]: The title and the first section could be simplified somehow.  I tried to do it in the following

----

# SQLFlow Command-line Client

## Motivations

At the moment, users can build SQLFlow into a gRPC server or an [command-line binary](../run/repl.md).  Both forms statically link to the SQLFlow core code and inherit all dependencies, including TensorFlow, various database client libraries, and various SQL dialect parsers in the form of gRPC backend servers.  The many dependencies require us to release and run both forms in Docker containers.

For convenient deployment, we want to run the SQLFlow server in containers, and end-users access the server via the command-line client.  In particular, we want it easy to install the client tool, so we need to prune the dependencies.  As a solution, we make the client tool a statically linked Go binary that remotely calls the SQLFlow server, so users just download and run it.

The command-line client tool, named `sqlflow`, is complementary to other forms of clients, like Jupyter Notebook.  Users can write shell or Python scripts calling the client tool to realize complex applications.

File: doc/design/cli.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+1. We don't need the flag `--model_dir` any more because the training process is running in `sqlflowserver`.
+
+1. We'll define a new flag `--host` that holds the `sqlflowserver` address.
+
+#### Configuration
+We still use `godotenv` as well as environment variables to hold flags that change infrequently. The two necessary config entries are:
+
+1. `SQLFLOW_DATASOURCE` the alternative of `--datasource`. We already have this.
+
+1. `SQLFLOW_SERVER_HOST` the alternative of `--host`. This is to be added.
```
Comment (wangkuiyi) [inline doc/design/cli.md]: SQLFLOW_SERVER_HOST => SQLFLOW_SERVER

File: doc/design/cli.md [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+1. Wrap the current SQLFlow REPL in a script, we already have such an [attempt](https://github.com/sql-machine-learning/sqlflow/pull/2114).
+1. Decouple the REPL implementation from the core code of SQLFlow and provide a native binary that only functions as a pure client.
+
+We decide to choose the second solution because it's more lightweight, if a user already has a SQLFlow deployment in her company, all she has to do is to download the native binary.
+
+## Design
+
+### Naming
+
+`REPL` is not a good name, We'll rename the binary as simply `sqlflow`.
```
Comment (wangkuiyi) [inline doc/design/cli.md]: We need more clear explanation here:

Currently, the command-line binary form of SQLFlow has the name `repl`, which is the name of a user interface design philosophy.  Many command-line tools, including `mysql` and `python`, implement the UI of REPL.  Let us follow the convention of `mysql` and `mysqld` to name the new client `sqlflow`.

File: doc/design/cli.md [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+### Dependencies and Installation
+
+The original `repl`'s only dependency is the `it2check` bash script that determines whether to call `sixel` to render images.
+
+We'll provide installation scripts on Mac (Bash) and Windows (PowerShell) to install `sqlflow` and its dependencies swiftly. The four scripts are:
```
Comment (wangkuiyi) [inline doc/design/cli.md]: I am afraid that this is too much for users. Let us make sure that users only need to download the client tool before running it. If necessary, let us abandon it2check and other cool stuff at the right moment.

File: doc/design/cli.md [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+1. [Table in protobuf](https://github.com/sql-machine-learning/sqlflow/blob/be7c5728f47e8d3b81893c1d712974a6ddcd5f1c/cmd/repl/repl.go#L161-L163)
+
+1. [The step package](https://github.com/sql-machine-learning/sqlflow/blob/be7c5728f47e8d3b81893c1d712974a6ddcd5f1c/cmd/repl/repl.go#L35)
+
+#### [Command Line Flags](https://github.com/sql-machine-learning/sqlflow/blob/be7c5728f47e8d3b81893c1d712974a6ddcd5f1c/cmd/repl/repl.go#L261-L268)
+
+1. We don't need the flag `--model_dir` any more because the training process is running in `sqlflowserver`.
+
+1. We'll define a new flag `--host` that holds the `sqlflowserver` address.
```
Comment (wangkuiyi) [inline doc/design/cli.md]: Not `--host`, but `--sqlflow_server`.  Let us make sure all names are accurate and precise.

File: doc/design/cli.md [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+1. Use the stub to send the SQL statement get from `ReadStmt` to the server.
+
+1. Parse the result returned from the SQLFlow server.
+    - For non-sixel terminal simulators, the server side should return `Plotille` ASCII figures. See https://github.com/sql-machine-learning/sqlflow/blob/be7c5728f47e8d3b81893c1d712974a6ddcd5f1c/pkg/step/step.go#L69-L85
+
+1. Both the client construction code and result parsing code can be implemented by referring to the existing code in `sqlflowserver/main_test.go`.
+
+#### Auto-complete
+The original implementation of auto-complete is based on the `pkg/sql/attribute` package, in a pure client as `sqlflow`, we can only depend on such packages partly because the client may have a different version with the server-side.
```
Comment (wangkuiyi) [inline doc/design/cli.md:73]: Let us remove this auto-completion feature at the first stage. We need a minimum viable product at the moment. Let us define it clearly in this design doc.

File: doc/design/cli.md [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+#### [RunSQLProgramAndPrintResult](https://github.com/sql-machine-learning/sqlflow/blob/be7c5728f47e8d3b81893c1d712974a6ddcd5f1c/cmd/repl/repl.go#L175)
+ 
+We have to replace the call to `step.RunSQLProgramAndPrintResult` with a new function `RunSQLProgram` that would be defined in the `sqlflow` codebase. The new function should:
+
+1. Construct a client stub from `--host`. 
+
+1. Use the stub to send the SQL statement get from `ReadStmt` to the server.
+
+1. Parse the result returned from the SQLFlow server.
+    - For non-sixel terminal simulators, the server side should return `Plotille` ASCII figures. See https://github.com/sql-machine-learning/sqlflow/blob/be7c5728f47e8d3b81893c1d712974a6ddcd5f1c/pkg/step/step.go#L69-L85
```
Comment (wangkuiyi) [inline doc/design/cli.md]: Plotille cannot draw images with high-precision.  Let us save the image to local files and optionally start a viewer program.  We can start Preview.app on macOS, or the `display` command-line tool from ImageMagick on Linux with X-window.

**Review by wangkuiyi — CHANGES_REQUESTED** [review]

File: doc/design/cli.md (hunk 1 above)
Comment (shendiaomo) [reply doc/design/cli.md:1]: Done.

File: doc/design/cli.md (hunk 2 above)
Comment (shendiaomo) [reply doc/design/cli.md]: Done.

File: doc/design/cli.md (hunk 3 above)
Comment (shendiaomo) [reply doc/design/cli.md]: Done.

File: doc/design/cli.md (hunk 6 above)
Comment (shendiaomo) [reply doc/design/cli.md:73]: Done.

File: doc/design/cli.md (hunk 7 above)
Comment (shendiaomo) [reply doc/design/cli.md]: The client won't rely on Plotille, it's a dependency of SQLFlow server. We only need to get the Plotille file from the server.

Anyway, we'll rethink how to show figures before integrating any related new features.

File: doc/design/cli.md (hunk 4 above)
Comment (shendiaomo) [reply doc/design/cli.md]: Done. 
The client will only provide auto-complete features for built-in models (canned estimators and XGBoost) at the 1st stage. 
The auto-complete features for `sqlflow_models` and Model Zoo models will be designed at the next state when necessary.

`it2check` has nothing to do with auto-completion, we'll avoid this dependency at the first stage

File: doc/design/cli.md (hunk 5 above)
Comment (shendiaomo) [reply doc/design/cli.md]: Done.

**Review by typhoonzero — CHANGES_REQUESTED** [review]
Can we just reuse the repo `pysqlflow` so that:

1. users can use `pip install sqlflow` to install the command line together with necessary dependencies.
1. `pysqlflow` is already a gRPC client that support sending SQL statements to the SQLFlow server, just add an interactive command line to it.

**Comment by shendiaomo** [comment]
> Can we just reuse the repo `pysqlflow` so that:
> 
> 1. users can use `pip install sqlflow` to install the command line together with necessary dependencies.
> 2. `pysqlflow` is already a gRPC client that support sending SQL statements to the SQLFlow server, just add an interactive command line to it.

At the moment, releasing `sqlflow` basing on `repl` is much more practical than basing on `pysqlflow` in three aspects:
1. The binary doesn't depend on any other software including python. 
2. Adding a gRPC client to `repl` is much less of work than adding a qualified interactive interface to `pysqlflow`.
3. We should provide our best to the users.

**Review by wangkuiyi — APPROVED** [review]
I agree that the client tool is in Go other than in Python.  A Go program can be statically-linked and has no dependencies at all. This makes the installation easier and to support more platforms.

**Review by wangkuiyi — APPROVED** [review]

**Review by typhoonzero — APPROVED** [review]

**Pull request merged**